### ISSUE TITLE
date detection in read_csv_auto() 

### ISSUE THREAD

**Issue by luxint** [cwida/duckdb#816]
Apparently the read_csv_auto() function only detect dates in YYYY-MM-DD format, since there is a lot of ambiguity in date formats would it be possible to add a date format specifier to this function, maybe like what is used in the strftime() function in C? 
Example:  read_csv_auto('myfile.csv',"%m/%d/%Y")

**Comment by Mytherin** [comment]
That's indeed a good idea. We will first need to add this function before we can consider adding it to the CSV reader though, some earlier discussion [here](https://github.com/cwida/duckdb/issues/600).

I had two more thoughts w.r.t. this:
(1) Perhaps it would be useful for the user to specify this setting system-wide, i.e. the user could say `SET date_format='%m/%d/%Y` and that format would be used for both date import and export by default.
(2) Once we have several date format parsers, the CSV auto reader can try several popular options to see which one would work. There is always the general problem here though that figuring out if `01/02/2020` is February 1st or January 2nd is impossible to figure out. The hope would be that there would be another date in the column of the CSV file along the lines of `01/20/2020` or `20/02/2020` that would give the parser some certainty there.

The parser could even throw an error in case of these ambiguities (e.g. `error: found two potential date formats for column X -> "%m/%d/%Y and %d/%m/%Y, please manually select one...`)

**Comment by luxint** [comment]
ad(1) My experience is that system wide settings don't really work in this case,  you could import csv files from different sources/countries. A US file will probably have m/d/Y, a German file dd.mm.YYYY and so on.
ad(2) yes you could do that, I actually do this when I preprocess a csv file. However, why waist cpu cycles when you know/can  easily see the date format which removes all ambiguity. Probably also a good idea to trim/remove whitespace first.

**Comment by Alex-Monahan** [comment]
If possible, it would be great to have an option to suppress that error and just make a best effort. We have DuckDB running as a part of an analytics web server, so it isn't very easy to make a modification and rerun.

**Comment by hawkfish** [comment]
The number of date formats in the world is truly terrifying.  See [this white paper](https://research.tableau.com/sites/default/files/TR-2018-01.pdf) for details.  And I agree that file formats can be orthogonal to system settings as in ad(1). But having a default might also make things easier for the common case.  

[Postgres' COPY command](https://www.postgresql.org/docs/12/sql-copy.html) uses a single format:

> COPY input and output is affected by DateStyle. To ensure portability to other PostgreSQL installations that might use non-default DateStyle settings, DateStyle should be set to ISO before using COPY TO. It is also a good idea to avoid dumping data with IntervalStyle set to sql_standard, because negative interval values might be misinterpreted by a server that has a different setting for IntervalStyle.

Anything more complex is really a data cleaning problem and should be out of scope.

**Comment by Mytherin** [comment]
Currently our behavior is to default to ISO-8601, allowing the user to specify a `DATEFORMAT` and `TIMESTAMPFORMAT` using strftime syntax ([see here](https://duckdb.org/docs/sql/functions/dateformat)). Perhaps that is sufficient then? We could add date/timestamp format detection but, as mentioned by @hawkfish, this is very difficult and maybe out of scope.

**Comment by hawkfish** [comment]
From reading other discussions of the CSV reader, it sounds like a sample is taken for schema inference. While the number of date formats can be large, you can also do a decent job of detecting the most common ones when processing the sample. So having a list of formats to check such as ISO, `DATEFORMAT`, US and European formats (2 and 4 digit years) should cover most data sets. Then when we add a date format parser, that can be an override. We will need to write fast parsers for the basic formats.

 I've done all this stuff and can take this on unless someone else is interested.

**Comment by Mytherin** [comment]
That would be fantastic! Thanks for offering :)

We already have a fast ISO-8601 parser; it would make a lot of sense to have fast parsers for the other common formats.

Tie-breaking with the users locale is an option, but we currently don't have a system-wide locale. I also prefer not to put any behavior in the system that changes based on system-locale. I think it might be cleaner to either throw an error (ambiguous dateformat), or fallback to converting to string. Perhaps that behavior could be a flag in the copy options as well.

The auto detection based on samples happens in BufferedCSVReader::SniffCSV, standard date parsing happens in Date::TryConvertDate, and user-specified formats are located in `src/function/scalar/date/strftime.cpp`. Those would be good starting points.

**Comment by hawkfish** [comment]
Ok. Probably be about two weeks before I can get to it what with VLDB and the US Labour Day weekend.

**Comment by Mytherin** [comment]
This has been implemented now by @hawkfish, thanks again :)
